### ISSUE TITLE
Add ability to access previous values & cancel updates

### ISSUE THREAD

**Issue by theochampion** [rrousselGit/river_pod#219]
Hey there 😃 

I'm migrating my app to river_pod, love it so far, but I'm always running into the same issue.

Most of my state is asynchronous and depends on some other asynchronous state (chain of providers watching each other), in other terms, expensive operations, so I need to avoid as many unnecessary updates as possible. 

The problem is that there is currently no way to cancel an update based on a condition in a provider, thus, an update in the first provider will trigger an update in every of its "child" provider, no matter if that update is necessary or not.

Now to a more concrete example:

In my app, I have a `changeNotifierProvider` that emits the user's location every few seconds. This provider is watched by many other providers in the app. One of them is a `FutureProvider` that, from the location, make a network call to reverse geocode it and get the city name. This provider is itself watched by another `FutureProvider` that, from that city name, make another network call to get the trending posts in the area.

so we have

`changeNotifierProvider` (`A`) (location)
                  |
v
      `FutureProvider` (`B`) (net call, get city name)
                  |
v
      `FutureProvider` (`C`) (net call, get trending posts in that city)
                  |
v
         `Consumer` (`D`) display list of trending posts

What happens here is, every time the location update (every few seconds), `B` makes a network call, once it is done, `C` receive the change and make a network call, finally `D` updates with the value. A lot of unnecessary calls and UI jitter (as `D` goes back to the loading state every few seconds).

What I would like to be able to do instead is, in `B` read the previously received value, here a location, and carry out the update OR NOT based on that. Here, the conditions could be:

in `B`
if the location is the same as the previous one -> no update
if the location is is geographicaly less than 500m from the previous one -> no update

in `C`
if the city is the same as previous -> no update

This would prevent all the unnecessary updates down the line.

**Describe the solution you'd like**

I don't have enough knowledge of the inner workings of river_pod to make a relevant input as to how it should be implemented but being able to "reduce" updates like that implies two things:

##### a controllable way to know about previous updates

here i'm suggesting the use of an accumulator similar to the reduce function https://api.dart.dev/be/155145/dart-core/Iterable/reduce.html

this way the users are free to store whatever data from the previous updates to make their decision about wether to update or not

##### the ability to cancel an update

Hard to say what would make the most sense here. Maybe returning the previous value ? Or some kind of lifecycle like in react  https://developmentarc.gitbooks.io/react-indepth/content/life_cycle/update/using_should_component_update.html

**Describe alternatives you've considered**

I haven't found a clean way to go around this problem.

Happy to discuss what you guys think,

Cheers ! ✨ ❤️

**Comment by davidmartos96** [comment]
I would like to add that I've seen other people's issues where they ask for a way to read the previous value as well in the `ProviderListener` callback. Although I'm not able to find them now. Maybe it was in the Discord.

**Comment by EdwynZN** [comment]
It would be really nice to have a way to check previous values to do some other logic (just as Provider has updateShouldNotifiy), right now the way to avoid unnecesary rebuilds is creating a Provider with the specific value you care and check on that to rebuild other Providers [My provider updates too often, what can I do?](https://riverpod.dev/docs/concepts/combining_providers#my-provider-updates-too-often-what-can-i-do)

In your case there are 2 options:
```
changeNotifierProvider (A) (location) 
// doing some logic just like ValueNotifier to check if the location value to replace is different from the previous
```

A provider between A and B like the example in the docs

```
final _locationProvider = Provider<Position>((ref) {
  return ref.watch(AProvider).position; // I don't know what library you're using for location
});
```

The problem with the second one is that you cannot see the previous value to check if they're different but still in the 500m radius proximity, which could be a really nice feature to have

**Comment by iamarnas** [comment]
@EdwynZN It's not easier just read or watch providers inside ChangeNotifier and tell when notifi. Like in this example.
```dart
final repositoryProvider = ChangeNotifierProvider((ref) => Repository(ref));

class Repository extends ChangeNotifier{
  Repository(this.ref);

  final ProviderReference ref;
}
```

**Comment by theochampion** [comment]
@iamarnas I'm not sure I understand what you are proposing. How does watch work from inside the ChangeNotifier ? Since there is no callback on watch, how do you update the internal state of the ChangeNotifier according to the watch ?

**Comment by iamarnas** [comment]
@theochampion I use it with `AsyncValue` when I handle state inside `StateNotifier` when I change watched provider everywhere in the context then `AsyncValue` automatically rebuild state. I write from my phone can't provide some example.

**Comment by rrousselGit** [comment]
Being able to access the previous value is "dangerous".

For things like ChangeNotifier/StateNotifier, the previous value is unusable because it was already disposed of.

There's also the question of "what value do we want to obtain?"
Because providers expose multiple values at one.

For example with a FutureProvider, the previous value is the previously created Future or the last emitted AsyncValue?

**Comment by theochampion** [comment]
@iamarnas So if I understand, what you are suggesting is doing what I ended up doing. I created another `StateProvider` between `A` and `B` and in my `ChangeNotifier` `A` I update the value of this `StateProvider` only if some conditions are met (distance > 500m, etc...). This works well but introduces the need for an intermediary provider as well as the need for `A` to handle some "foreign" logic. @rrousselGit is that the river_pod way to handle such problem ?

@rrousselGit To answer the question "what value do we want to obtain?". I suggest the use of an accumulator. That way the developer is responsible to add (or not) the values he will need for the next update. The accumulator could also make sure the data is still available after a dispose.

**Comment by rrousselGit** [comment]
> I suggest the use of an accumulator

Do you have an example? I don't quite understand how that would allow a StreamProvider to obtain the previous stream vs the last emitted value

**Comment by rrousselGit** [comment]
As I come back to this:

I believe an alternate way of listening to providers would solve this:


```dart
final first = SomeProvider(...);

final second = ChangeNotifierProvider((ref) {
  final notifier = MyNotifier();

  listen<Something>(first, (something) {
    notifier.somethingChanged(something);
  });
});
```

**Comment by isanimu** [comment]
@theochampion What you really need is a combiner/reducer solution like [RxDart](https://github.com/ReactiveX/rxdart).

First, just like others already mention, you shouldn't break your data source/logic into multiple providers. Instead you should use a repository pattern and put all your logic inside there.

> in B
if the location is the same as the previous one -> no update
if the location is is geographicaly less than 500m from the previous one -> no update

```dart
stream.distinct()
```

Distinct operator is basically what you need if you need to filter between previous and current stream value. It must return true to be _**distincted**_.

```dart
stream.distinct((previous, current) => isEqual(previous, current) && isLessThan500m(previous, current));
```

If you need basic filter you can use Where operator.

```dart
stream.where((location) => location.iWantThis());
```

> in C
if the city is the same as previous -> no update

I'm assuming that case should already been solved using above solution.

Then, once you already got the value that you needed, you can use SwitchMap operator to perform fetching data and transform it into a view state for your ui.

```dart
stream
  .distinct()
  .switchMap<ViewState>(getTrendingPosts)
  .startWith(InitialState())

Stream<ViewState> getTrendingPosts(Location location) async* {
  yield LoadingState();
  notifyListeners();

  yield LoadedState(await api.fetchTrendingPosts(location));
  notifyListeners();
}
```

So, something like this could work.

main.dart
```dart
import 'dart:math';
import 'package:flutter/material.dart';
import 'package:equatable/equatable.dart';
import 'package:hooks_riverpod/hooks_riverpod.dart';
import 'package:rxdart/rxdart.dart';

void main() {
  runApp(ProviderScope(child: MyApp()));
}

class MyApp extends StatelessWidget {
  const MyApp({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Home(),
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
    );
  }
}

class Location extends Equatable {
  final int value;

  Location(this.value);

  static bool isEqual(Location previous, Location current) {
    return (current.value == previous.value);
  }

  static bool isLessThan500m(Location previous, Location current) {
    return current.value < 3;
  }

  static bool distinct(Location previous, Location current) {
    return isEqual(previous, current) && isLessThan500m(previous, current);
  }

  @override
  List<Object> get props => [value];
}

class Post {
  final String title;

  Post({this.title});
}

abstract class ViewState extends Equatable {}

class InitialState extends ViewState {
  @override
  List<Object> get props => [];
}

class LoadingState extends ViewState {
  @override
  List<Object> get props => [];
}

class LoadedState extends ViewState {
  final List<Post> posts;
  LoadedState(this.posts);

  @override
  List<Object> get props => [posts];
}

class ErrorState extends ViewState {
  @override
  List<Object> get props => [];
}

class Api {
  Future<List<Post>> fetchTrendingPosts(Location location) async {
    await Future.delayed(Duration(milliseconds: 250));
    print('onApi - ${location.value}');
    return [
      Post(title: 'Hello World: ${location.value}'),
    ];
  }
}

final apiProvider = Provider<Api>((_) => Api());

class TrendingNotifier extends ChangeNotifier {
  final Api _api;
  final PublishSubject<Location> _sink = PublishSubject<Location>();
  Stream<ViewState> _state;

  TrendingNotifier(ProviderReference ref) : _api = ref.read(apiProvider) {
    _state = _sink
        .distinct(Location.distinct)
        .switchMap<ViewState>(getTrendingPosts)
        .startWith(InitialState());
  }

  Stream<ViewState> get state => _state;

  void updateLocation(Location location) {
    _sink.add(location);
  }

  Stream<ViewState> getTrendingPosts(Location location) async* {
    yield LoadingState();
    notifyListeners();
    await Future.delayed(Duration(milliseconds: 250));

    try {
      final posts = await _api.fetchTrendingPosts(location);
      yield LoadedState(posts);
    } catch (e) {
      yield ErrorState();
    }
    notifyListeners();
  }

  @override
  void dispose() {
    _sink.close();
    super.dispose();
  }
}

final sinkProvider =
    Provider<TrendingNotifier>((ref) => TrendingNotifier(ref));
final stateProvider = StreamProvider<ViewState>(
    (ref) => ref.read(sinkProvider).state);

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Consumer(
                builder: (context, watch, _) {
                  final provider = watch(stateProvider);
                  final state = provider.data?.value;
                  if (state is LoadingState) return CircularProgressIndicator();
                  if (state is LoadedState) return Text(state.posts[0].title);
                  return Text('Click the Refresh button!');
                },
              )
            ],
          ),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: () {
            context
                .read(sinkProvider)
                .updateLocation(Location(Random().nextInt(5)));
          },
          tooltip: 'Refresh',
          child: Icon(Icons.refresh),
        ));
  }
}
```

pubspec.yaml
```
dependencies:
  flutter:
    sdk: flutter
  flutter_riverpod: ^0.12.1
  rxdart: ^0.25.0
  equatable: ^1.2.5
```

**Comment by rrousselGit** [comment]
No, you shouldn't need rxdart.

One of the goals of riverpod is to **not** use RX, because it has a very complex API

**Comment by rrousselGit** [comment]
Closing in favor of #335 which should fix this issue by both supporting `ref.listen` and `ref.watch(provider.select)` to offer a better control over when a provider "rebuilds"
